### PR TITLE
Add methods for path type.

### PR DESCRIPTION
**Pull request by CPWstatic**

### CODE REVIEW

File: src/common/datatypes/Path.cpp [hunk 1]
```diff
@@ -7,5 +7,28 @@
...
+void Path::reverse() {
+    if (steps.empty()) {
+        return;
+    }
+    std::reverse(steps.begin(), steps.end());
+    swap(src, steps.front().dst);
+    for (size_t i = 0; i < steps.size(); ++i) {
+        if ((i + 2) <= steps.size()) {
+            swap(steps[i].dst, steps[i + 1].dst);
+        }
```
Comment (dutor) [inline src/common/datatypes/Path.cpp:19]: `if ( i + 1 < steps.size())` is more clear and consistent with the loop condition and swap.

Comment (dutor) [reply src/common/datatypes/Path.cpp:19]: BTW. If the condition is `i < steps.size() - 1`, I think you could remove the `if`, and reverse the last edge after the loop.

File: src/common/datatypes/Path.cpp [hunk 2]
```diff
@@ -7,5 +7,28 @@
...
+        steps[i].type = -steps[i].type;
+    }
+}
 
+bool Path::append(Path path) {
+    if (src != path.src && steps.back().dst != path.src) {
+        return false;
+    }
```
Comment (dutor) [inline src/common/datatypes/Path.cpp:27]: Better add some comments about the semantics of this interface.

**Review by dutor — DISMISSED** [review]
LGTM.

File: src/common/datatypes/Path.cpp [hunk 3]
```diff
@@ -7,5 +7,27 @@
...
+        steps[i].type = -steps[i].type;
+    }
+    steps.back().type = -steps.back().type;
+}
 
+bool Path::append(Path path) {
+    if (src != path.src && steps.back().dst != path.src) {
+        return false;
+    }
+    steps.reserve(path.steps.size());
```
Comment (Shylock-Hg) [inline src/common/datatypes/Path.cpp:27]: steps.size() + path.steps.size()

Comment (CPWstatic) [reply src/common/datatypes/Path.cpp:27]: The `reserve` is about to increase the capacity.

Comment (Shylock-Hg) [reply src/common/datatypes/Path.cpp:27]: The parameter is the new capacity instead of increase capacity.

**Review by laura-ding — APPROVED** [review]

File: src/common/datatypes/Path.cpp (hunk 3 above)
Comment (yixinglu) [reply src/common/datatypes/Path.cpp:27]: steps.size() + path.steps.size() - 1

**Review by Shylock-Hg — APPROVED** [review]

**Review by dutor — APPROVED** [review]

**Pull request merged**